File: 40-combination-sum-ii/combination-sum-ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] cand, int target) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(cand);
        helper(new ArrayList<>(), 0, target, 0, cand, res);
        return res;
    }

    private void helper(List<Integer> sub, int sum, int target, int index, int[] cand, List<List<Integer>> res) {
        if (sum == target) {
            res.add(new ArrayList<>(sub));
            return;
        }

        if (sum > target || index == cand.length) {
            return;
        }

        // pick
        sub.add(cand[index]);
        helper(sub, sum + cand[index], target, index + 1, cand, res);
        sub.remove(sub.size() - 1);
        
        while (index < cand.length - 1 && cand[index + 1] == cand[index]) {
            index++;
        }
        helper(sub, sum, target, index + 1, cand, res);
    }
}

File: 40-combination-sum-ii/combination-sum-ii.cpp
class Solution {
public:
    void help(int id, int sum, int target, vector<int> &candidates, vector<int> &cur, vector<vector<int>> &ot) {
        if (sum == target) {
            ot.push_back(cur);
            return;
        }
        if (id >= candidates.size() || sum > target) return;

        for (int i = id; i < candidates.size(); i++) {
            if (i > id && candidates[i] == candidates[i - 1]) continue;

            cur.push_back(candidates[i]);
            help(i + 1, sum + candidates[i], target, candidates, cur, ot);
            cur.pop_back();
        }
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<int> cur;
        vector<vector<int>> ot;
        help(0, 0, target, candidates, cur, ot);
        return ot;
    }
};
